crates/ty_python_semantic/src/dunder_all.rs

```diff
@@ -0,0 +1,469 @@
+    file: File,
+
+    /// The scope in which the `__all__` names are being collected from.
+    ///
+    /// This is always going to be the global scope of the module.
+    scope: ScopeId<'db>,
+
+    /// The semantic index for the module.
+    index: &'db SemanticIndex<'db>,
+
+    /// The origin of the `__all__` variable in the current module, [`None`] if it is not defined.
+    origin: Option<DunderAllOrigin>,
+
+    /// A flag indicating whether the module uses unrecognized `__all__` idioms or there are any
+    /// invalid elements in `__all__`.
+    invalid: bool,
+
+    /// A set of names found in `__all__` for the current module.
+    names: FxHashSet<Name>,
+}
+
+impl<'db> DunderAllNamesCollector<'db> {
+    fn new(db: &'db dyn Db, file: File, index: &'db SemanticIndex<'db>) -> Self {
+        Self {
+            db,
+            file,
+            scope: global_scope(db, file),
+            index,
+            origin: None,
+            invalid: false,
+            names: FxHashSet::default(),
+        }
+    }
+
+    /// Updates the origin of `__all__` in the current module.
+    ///
+    /// This will clear existing names if the origin is changed to mimic the behavior of overriding
+    /// `__all__` in the current module.
+    fn update_origin(&mut self, origin: DunderAllOrigin) {
+        if self.origin.is_some() {
+            self.names.clear();
+        }
+        self.origin = Some(origin);
+    }
+
+    /// Extends the current set of names with the names from the given expression which can be
+    /// either a list of names or a submodule's `__all__` variable.
+    ///
+    /// Returns `true` if the expression is a valid list or submodule `__all__`, `false` otherwise.
+    fn extend_from_list_or_submodule(&mut self, expr: &ast::Expr) -> bool {
+        match expr {
+            // `__all__ += [...]`
+            // `__all__.extend([...])`
+            ast::Expr::List(ast::ExprList { elts, .. }) => self.add_names(elts),
+
+            // `__all__ += submodule.__all__`
+            // `__all__.extend(submodule.__all__)`
+            ast::Expr::Attribute(ast::ExprAttribute { value, attr, .. }) => {
+                if attr != "__all__" {
+                    return false;
+                }
+                let Some(name_node) = value.as_name_expr() else {
+                    return false;
+                };
+                let Symbol::Type(ty, Boundness::Bound) = symbol_from_bindings(
+                    self.db,
+                    self.index
+                        .use_def_map(self.scope.file_scope_id(self.db))
+                        .bindings_at_use(name_node.scoped_use_id(self.db, self.scope)),
+                ) else {
+                    return false;
+                };
+                let Some(module_literal) = ty.into_module_literal() else {
+                    return false;
+                };
+                // TODO: Do we need to check if the module is a submodule of the current module?
```

Comment:
I assume this TODO comes from the spec's use of `submodule.__all__` in its example? I don't think this is a clear specification that it must be a submodule, and I don't really see any advantage to that requirement. In fact I think typeshed does this in non-submodule cases (`collections.abc` reuses the `__all__` from `_collections_abc`, which is not a submodule -- I guess this isn't "extending", but it doesn't make sense to have different rules for which external `__all__` can be used.)

Reply:
Yeah, it comes from the usage of "submodule" and I wasn't sure about whether that's a requirement or not which is why I marked it as TODO.

That makes sense. I'll remove the TODO and update all usages of "submodule" to just use "module" instead.